Feat/transient

nobiot opened the pull request:
WIP. Creating this PR to make the work visible.

nobiot committed:
refactor: Move org-transclusion-add-payload

nobiot committed:
feat: "auto" level property #195 #254

nobiot committed:
feat: WIP no-first-heading #254

nobiot committed:
feat: add org-transient

nobiot committed:
add(transient): options from src-lines extension

nobiot committed:
feat(transient): insert-from-link

nobiot committed:
wip(transient): add Setting to control visibility

nobiot committed:
chg: Make hooks to set up extensions removable #261

Added the  following utility functions to support the move

- `org-transclusion-extension-functions-add-or-remove`
- `org-transclusion-extension-set-a-hook-functions`

Modified function `org-transclusion-load-extensions-maybe` so that setting
`org-transclusion-extensions` with the `customize` facilities will automatically
activate the corresponding minor mode for each extension.

Use of `customize` and `org-transclusion-extensions` is optional. Users can
ignore these and use extensions by loading and activating minor modes in their
init.el file.

So, this below should work (each minor mode function has the autoload cookie, so
the library should not have to be explicitly loaded).

```
(with-eval-after-load 'org-transclusion
    (org-transclusion-indent-mode +1))
 ```

Adding to the list and requiring the library like this below did not use to be
necessary...

```
(with-eval-after-load 'org-transclusion
     (add-to-list 'org-transclusion-extensions 'org-transclusion-indent-mode)
     (require 'org-transclusion-indent-mode))
```

The Customizing facility would load the active extension library with
`require`.... But I think this was never intuitive.

nobiot committed:
style: No need cl-macs (no cl-loop used in the end)

nobiot committed:
fix: org-transclusion-*-extension-functions

'((cons e1 e2)) is incorrect. `cons` will be part of the list.

nobiot committed:
chg: Add support Org-id with search option `[[id:<id>::heading title]]`

nobiot committed:
chg(org-id): Generalize org-id-open by using org-link-open instead

nobiot committed:
fix(add): org-id with search option (::*heading) #237

While fixing the issue, I am extending the "add" operation to allow for *any*
link.  A lot is still under development and many things will change. But this
seems to work without breaking any user-facing commands and function.

Introduced a new function `org-transclusion-add-target-marker`, which helps
transcluding *any* link (as long as Org can reach the link target and Emacs
creates the link's target buffer -- tested with `notmuch` link successfully).

Refactored the code for the "add" operation with this new approach.

nobiot committed:
chg: fn signature change,  org-transclusion-content-format-functions

The third (last) argument was INDENT, which is only a property within
KEYWORD-VALUES (KEYWORD-PLIST).

Changed to the latter to make the function more flexible.

nobiot committed:
chg(WIP): Remove org-transclusion-add-payload

This is a beginning of a large refactor for the whole `add` operation.

nobiot committed:
chg: Refactor & fix auto level and remove first heading

nobiot committed:
refactor: Remove old functions that were commented out

nobiot committed:
chg(transient): Menu outputs user-error if used in non-Org buffer

okomestudio commented:
I'm adding this comment in the context of https://github.com/nobiot/org-transclusion/issues/271. (Thanks for letting me try this dev branch!)

When transcluding an ID link with a dedicated target/search (i.e., `[[id:<id>::<target>][<desc>]]`), the logic in the following line doesn't work as intended:

https://github.com/nobiot/org-transclusion/blob/feat/transient/org-transclusion.el#L974-L978

when the target is in the section before the first heading (in which case the entire document will be transcluded). I'm not exactly sure the best way to detect whether the transcluded content is non-heading element, but the minimal check along the line of the following seems necessary:

``` emacs-lisp
               (if (and (org-before-first-heading-p)
			(not (looking-at-p "\\(#\\+name:\\|<<\\)")))
                   (org-transclusion-content-org-filtered
                    nil plist)
                 (org-transclusion-content-org-filtered
                  'only-element plist))
```

I am just reporting a bug in case this helps when you have time getting back to this branch. Otherwise, the feature branch has been very useful to me, and will continue use and test.

Thank you again for all the work.

nobiot commented:
@okomestudio Thank you for this. I think I understand the issue. Let me try a couple ideas…

k4r4b3y commented:
@nobiot I have been testing this branch.  I noticed that `id:<id>::<target>` linking doesn't work for the **same-file** targets.  Here's a simple example:

```
* 08:49 Example of learning human languages via anki and youtube :german:language:journals:2025_09_25:
:PROPERTIES:
:ID: 0c59c5a8-ffb4-40e3-b107-8d7893c2ce4d
:CREATED: [2025-09-25 Thu 08:49]
:END:

The [[https://cjauvin.github.io/posts/learning-persian/][blog post]] is short and simple.  However, showcases an effective
method of using youtube and spaced repetition with langauage learning.
The gist of it is that <<gist>>

* 08:53 Showcasing in-file targeting with org-transclusion
:PROPERTIES:
:ID: d697a00d-62b9-4d49-8196-b1ebadfdccda
:CREATED: [2025-09-25 Thu 08:53]
:END:

See this one:

#+transclude: [[id:0c59c5a8-ffb4-40e3-b107-8d7893c2ce4d::gist]]

or this one:

#+transclude: [[id:ccf3c642-fbef-4485-a21b-e83784c3303f::test]]

```

hitting `<f12>` (ie, `org-transclusion-add`) on the first `#+transclude:` statement works erroneously, while the second `#+transclude:` statement works correctly.  The former one targets the `id` that's on the same file (this is a org-roam-daily file) while the latter targets the `id` that's on the previous day's org-roam-daily file.

While the second transclusion works fine, the real problem is that the first transclusion wrongly inserts the `<<gist>>` up to the target id and gives me something like this which is totally false:

<img width="1063" height="612" alt="ss1" src="https://github.com/user-attachments/assets/fa20875f-469b-40b1-9074-5cc25149193f" />

Observe that the transcluded content is inserted right under the source content, which is a real headscratcher.

nobiot committed:
fix: ID links with search option (9.7+)

nobiot committed:
fix id:<id>::<target> linking in the target same buffer #273

This fixes the issue reported by k4r4b3y (on Sep 25)

    I have been testing this branch. I noticed that id:<id>::<target> linking
    doesn't work for the same-file targets.

In addition, this improves  function `org-transclusion-add-target-marker`

1. It now issues better error message when the link does not work

2. It now does not re-configure the buffer and windows, which was disrupting for
   end users

nobiot committed:
Merge remote-tracking branch 'origin/feat/transient'

nobiot committed:
fix(wip) org-transclusion-content-add-text-props-and-overlay

Syntax error has been fixed but the overlay/fringe do not seem to work properly
for `org-modern` on my end.

nobiot committed:
fix: Put back previous feature to allow for asynchronous transclusion

This is for backward compatibility. `hyperdrive-org-transclusion' requires this
payload function. I am going to ask the author to review. Eventually,
`org-transclusion-add-payload' needs to be removed.

nobiot committed:
refac: Remove unused let-vars - compiliation warning

cur-buf and cur-marker are unused.

nobiot committed:
docs: Updated modified date and version

nobiot committed:
docs: Updated copyright years

nobiot committed:
test: Fix noweb unit test file

nobiot committed:
fix(font-lock): font-lock-add-keywords to be comp with org-modern

The previous way that uses org-font-lock-extra-keywords gets overridden by
org-modern. font-lock-add-keywords works with it.

nobiot committed:
refac: org-transclusion-font-lock can be activated independently

 By default, org-transclusion-font-lock does not need to be explicitly
 configured as it will be automatically activated via
 `org-transclusion-extensions'. `org-transclusion-font-lock' is set to be active
 by default).

 Alternatively, you can set it up without loading the whole `org-transclusion'
 library.

 (use-package org-transclusion-font-lock
  :after org
  :config (org-transclusion-font-lock-mode +1))

nobiot committed:
fix: indenting can lead to duplicating transclusion when remove

See the code. To repro the issue, set `org-adapt-indentation' to t, and have a
\#+transclude: keyword in an indented paragraph. Then have two transclusions
side-by-side without space, add both of them, and remove the first one. You will
see that it is not fully removed.

nobiot committed:
feat: transient-menu and org-transclusion-insert

nobiot committed:
chg(transient): Adding more properties (:noweb-chunk, etc.)

nobiot committed:
fix(src): Use the new org-transclusion-add-target-marker

nobiot committed:
refac(src): search-option ::number and ::/regexp/ are limitations

Now in theory both work (and are working), only ::number makes sense.
::/regepxp/ opens an occur buffer for users to interact, which requires user
interaction. The source buffer is open behind the occur buffer but the point
does not move. Thus... ::/regexps/ is no longer a limitation; it's functionally
working but Org-transclusion does not interact with the occur buffer.

nobiot committed:
refact: Rename add-target-marker add-source-marker

nobiot committed:
refac: remove org-transclusion-fix-common-misspelling

This was a workaround when the cause of the infinite loop problem was not
identified (#177). It has been fixed with commit f6fd666. This workaround can be
safely removed (it had no effect).

nobiot committed:
refac: remove org-transclusion-fix-common-misspelling

This was a workaround when the cause of the infinite loop problem was not
identified (#177). It has been fixed with commit f6fd666. This workaround can be
safely removed (it had no effect).

nobiot committed:
fix: errornously indent following part when remove

nobiot committed:
fix(lines): Wrong indentation when org ID combined with :lines

nobiot committed:
fix: incorrect indent when no-first-heading and only-contents

Moved the following piece:

;; Fix indentation when `org-adapt-indentation' is non-nil
(org-indent-region beg end-mkr)

nobiot committed:
test: Updated test2.0.org file

nobiot committed:
test: updte files

nobiot committed:
fix(fringe): removing fringe does not work #283

Fixing non-Org buffers

nobiot committed:
test: file update

nobiot committed:
fix(fringe): removing fringe does not work #283

Fixing Org buffers

nobiot committed:
feat(insert): org-transclusion-insert-link-functions as a user option

nobiot committed:
fix: removing transclusions should not alter buffer-modified-p flag

meedstrom commented:
@nobiot Thanks for your work!!! Now org-node ships a pretty reliable command:

```elisp
(defun org-node-insert-transclusion-as-subtree (&optional node arg)
  "Insert a subheading, link, newline and transclusion.
Prompt for NODE if needed.
Result will basically look like:

** [[Note]]
#+transclude: [[Note]] :level auto :no-first-heading

Tip: If you often transclude file-level nodes with no initial heading,
consider configuring `org-transclusion-exclude-elements' to exclude
keywords.

Prefix argument ARG as in `org-insert-subheading'."
  (interactive "*i\nP" org-mode)
  (org-node-cache-ensure)
  (let (input)
    (unless node
      (setq input (org-node-read-candidate "Transclude node content: " t))
      (when (string-blank-p input)
        (setq input (funcall org-node-blank-input-title-generator)))
      (setq node (gethash input org-node--candidate<>entry)))
    (let ((id (if node (org-mem-id node) (org-id-new)))
          (title (if node (org-mem-title node) input))
          (buf (current-buffer))
          (pt (point-marker)))
      (unless node
        (setq node (org-node--create-sync title id)))
      (with-current-buffer buf
        (goto-char pt)
        (set-marker pt nil)
        ;; Never bring along any text from the current line.
        (org-node--safe-ensure-blank-line)
        (org-insert-subheading arg)
        (let* ((desc (or (and org-node-custom-link-format-fn
                              node
                              (funcall org-node-custom-link-format-fn node))
                         title))
               (link (org-link-make-string (concat "id:" id) desc)))
          (insert link)
          (save-excursion
            (newline-and-indent)
            (insert "#+transclude: " link " :level auto :no-first-heading"))
          (run-hooks 'org-node-insert-link-hook))))))
```
source https://github.com/meedstrom/org-node/blob/main/org-node.el#L1628

and now I don't need to worry about moving headings around, due to `:level auto`.  I'm happy with it :)
